File: packages/web/src/actions/apply_transaction.ts
import { set } from "~/stores/data";
import create_transaction from "./create_transaction";

export default function apply_transaction(tr: ReturnType<typeof create_transaction>) {
	console.log(JSON.stringify(tr, null, 2))
  set((draft) => {
    for (const item of tr.payload) {
      switch (item.type) {
        case "create_block": {
          draft.blocks[item.payload.id] = item.payload;
          break;
        }
        case "update_block": {
          if (draft.blocks[item.payload.id]) {
            // @ts-ignore
            draft.blocks[item.payload.id] = { ...draft.blocks[item.payload.id], ...item.payload };
          }
          break;
        }
        default: {
          // @ts-ignore
          console.error(`Unsupported transaction item type (${item.type})`);
        }
      }
    }
  });

  return tr;
}

File: internal/action/apply_transaction.go
package action

import (
	"context"
	"encoding/json"

	"github.com/immernote/immernote/internal/database"
)

type Transaction struct {
	Type    string `json:"type"`
	Payload []struct {
		Type    string           `json:"type"`
		Payload *json.RawMessage `json:"payload"`
	} `json:"payload"`
}

func ApplyTransaction(tr Transaction) error {
	tx, err := database.Get().Begin(context.Background())
	if err != nil {
		return err
	}

	for _, item := range tr.Payload {
		switch item.Type {
		case "create_block":
			params := new(CreateBlockParams)
			if err := json.Unmarshal(*item.Payload, params); err != nil {
				return err
			}
			if err := CreateBlock(*params, tx); err != nil {
				return err
			}
		case "update_block":
			params := new(UpdateBlockParams)
			if err := json.Unmarshal(*item.Payload, params); err != nil {
				return err
			}
			if err := UpdateBlock(*params, tx); err != nil {
				return err
			}
		}
	}

	if err := tx.Commit(context.Background()); err != nil {
		return err
	}

	return nil
}
